opensquirrel/instruction_library.py

```diff
@@ -3,18 +3,23 @@
 from collections.abc import Callable, Mapping
 from typing import TYPE_CHECKING
 
+from opensquirrel.default_instructions import default_gate_set, default_non_unitary_set
+from opensquirrel.utils import Singleton
+
 if TYPE_CHECKING:
     from opensquirrel.ir import Gate, NonUnitary
 
 
-class InstructionLibrary:
-    def __init__(
-        self,
-        gate_set: Mapping[str, Callable[..., Gate]],
-        non_unitary_set: Mapping[str, Callable[..., NonUnitary]],
-    ) -> None:
-        self.gate_set = gate_set
-        self.non_unitary_set = non_unitary_set
+class InstructionLibrary(metaclass=Singleton):
+    def __init__(self) -> None:
+        self.gate_set = default_gate_set
+        self.non_unitary_set = default_non_unitary_set
+
+    def get_gate_set(self) -> Mapping[str, Callable[..., Gate]]:
+        return self.gate_set
+
+    def get_non_unitary_set(self) -> Mapping[str, Callable[..., NonUnitary]]:
+        return self.non_unitary_set
```

Comment:
Could these be set up as properties?
```suggestion
    def __init__(self) -> None:
        self._gate_set = default_gate_set
        self._non_unitary_set = default_non_unitary_set
        
    @property    
    def gate_set(self) -> Mapping[str, Callable[..., Gate]]:
        return self._gate_set
    
    @gate_set.setter
    def gate_set(self, value: Mapping[str, Callable[..., Gate]]) -> None:
        self._gate_set = value

    @property
    def non_unitary_set(self) -> Mapping[str, Callable[..., NonUnitary]]:
        return self._non_unitary_set
    
    @gate_set.setter
    def non_unitary_set(self, value: Mapping[str, Callable[..., NonUnitary]]) -> None:
        self._non_unitary_set= value
```

Reply:
Hm... I don't really want to offer a setter.

What would be the benefit then to use a property? Just that you would use `InstructionLibrary().gate_set` instead of `InstructionLibrary.get_gate_set()`?

Reply:
That would indeed be the benefit. But it is also generally the way to implement getter functionality in Python.

You don't need to offer a setter that's fine.

Reply:
Done. Without settters.